Remove calls to free()

`free()` is not required in Go because it has automatic memory management. However, this relates to a larger issue of having code removed, rather than just transpiled directly.

Now, after transpile next C code:
```c
#include <stdlib.h>

int main() {
        int * buf1, *buf2, *buf3;
        buf1 = (int*) malloc(100*sizeof(int));
        buf2 = (int*) calloc(100,sizeof(int));
        buf3 = (int*) realloc(buf2, 500*sizeof(int));
        free(buf1);
        free(buf3);
        return 0;
}
```
we have errors:
```cmd
go run free.go
# command-line-arguments
./free.go:199:14: undefined: noarch.IntSliceToByteSlice
./free.go:200:14: undefined: noarch.IntSliceToByteSlice
./free.go:204:2: undefined: os
./free.go:206:14: undefined: os
```
Go code:
```golang
func main() {
        __init()
        var buf1 []int
        var buf2 []int
        var buf3 []int
        buf1 = make([]int, 100*4/4)
        buf2 = make([]int, 100*4/4)
        buf3 = make([]int, 500*4/4)
        noarch.Free(noarch.IntSliceToByteSlice(buf1))
        noarch.Free(noarch.IntSliceToByteSlice(buf3))
        return
}
```

--- a/program/function_definition.go
+++ b/program/function_definition.go
@@ -195,6 +195,7 @@ var builtInFunctionDefinitions = []string{
 	"long long strtoll(const char *, char **, int) -> noarch.Strtoll",
 	"long unsigned int strtoul(const char *, char **, int) -> noarch.Strtoul",
 	"long long unsigned int strtoull(const char *, char **, int) -> noarch.Strtoull",
+	"void free(void*) -> _",
 
 	// I'm not sure which header file these comes from?
 	"uint32 __builtin_bswap32(uint32) -> darwin.BSwap32",

--- a/program/import.go
+++ b/program/import.go
@@ -15,6 +15,10 @@ func (p *Program) Imports() []string {
 func (p *Program) AddImport(importPath string) {
 	quotedImportPath := strconv.Quote(importPath)
 
+	if len(importPath) == 0 {
+		return
+	}
+
 	for _, i := range p.imports {
 		if i == quotedImportPath {
 			// Already imported, ignore.

--- a/tests/stdlib.c
+++ b/tests/stdlib.c
@@ -152,9 +152,29 @@ void test_calloc()
     is_eq(d[4], 456);
 }
 
+void test_free()
+{
+	diag("free");
+  int * buffer1, * buffer2, * buffer3;
+  buffer1 = (int*) malloc (100*sizeof(int));
+  buffer2 = (int*) calloc (100,sizeof(int));
+  buffer3 = (int*) realloc (buffer2,500*sizeof(int));
+  int i = 0;
+  free ((i += 1, buffer1));
+  if (buffer2 != NULL){
+  	i+=1;
+  }
+  if (i == 2)
+  {
+  	free (buffer3);
+	i++;
+  }
+  is_eq(i,3);
+}
+
 int main()
 {
-    plan(741);
+    plan(742);
 
     char *endptr;
 
@@ -255,6 +275,8 @@ int main()
     // exit() is handled in tests/exit.c
 
     // free() is handled with the malloc and calloc tests.
+	diag("free");
+	test_free();
 
     diag("getenv")
     is_not_null(getenv("PATH"));

--- a/tests/stdlib.c
+++ b/tests/stdlib.c
@@ -154,22 +154,21 @@ void test_calloc()
 
 void test_free()
 {
-	diag("free");
-  int * buffer1, * buffer2, * buffer3;
-  buffer1 = (int*) malloc (100*sizeof(int));
-  buffer2 = (int*) calloc (100,sizeof(int));
-  buffer3 = (int*) realloc (buffer2,500*sizeof(int));
-  int i = 0;
-  free ((i += 1, buffer1));
-  if (buffer2 != NULL){
-  	i+=1;
-  }
-  if (i == 2)
-  {
-  	free (buffer3);
-	i++;
-  }
-  is_eq(i,3);
+	int * buffer1, * buffer2, * buffer3;
+	buffer1 = (int*) malloc (100*sizeof(int));
+	buffer2 = (int*) calloc (100,sizeof(int));
+	buffer3 = (int*) realloc (buffer2,500*sizeof(int));
+	int i = 0;
+	free ((i += 1, buffer1));
+	if (buffer2 != NULL){
+		i+=1;
+	}
+	if (i == 2)
+	{
+		free (buffer3);
+	  i++;
+	}
+	is_eq(i,3);
 }
 
 int main()

--- a/transpiler/call.go
+++ b/transpiler/call.go
@@ -204,6 +204,49 @@ func transpileCallExpr(n *ast.CallExpr, p *program.Program) (
 		}
 	}
 
+	// Added for support removing function `free` of <stdlib.h>
+	// Example of C code:
+	// free(i+=4,buffer)
+	// Example of result Go code:
+	// i += 4
+	// _ = buffer
+	if functionName == "_" && functionDef.Name == "free" && functionDef.Substitution == "_" {
+		var argName string
+		if v, ok := realArgs[0].(*goast.CallExpr); ok && len(realArgs) == 1 {
+			if vv, ok := v.Args[0].(*goast.Ident); ok && len(v.Args) == 1 {
+				argName = vv.Name
+			}
+			if vv, ok := v.Args[0].(*goast.ParenExpr); ok && len(v.Args) == 1 {
+				argName = vv.X.(*goast.Ident).Name
+			}
+		}
+		if v, ok := realArgs[0].(*goast.Ident); ok {
+			argName = v.Name
+		}
+		if argName != "" {
+			devNull := &goast.AssignStmt{
+				Lhs: []goast.Expr{
+					&goast.Ident{
+						Name: "_",
+					},
+				},
+				Tok: token.ASSIGN,
+				Rhs: []goast.Expr{
+					&goast.Ident{
+						Name: argName,
+					},
+				},
+			}
+			preStmts = append(preStmts, devNull)
+			return nil, "", preStmts, postStmts, nil
+		}
+	}
+
+	if functionName == "_" {
+		errorMessage := fmt.Sprintf("function with name: %s , cannot be used in Go", functionName)
+		p.AddMessage(p.GenerateWarningMessage(errors.New(errorMessage), n))
+	}
+
 	return util.NewCallExpr(functionName, realArgs...),
 		functionDef.ReturnType, preStmts, postStmts, nil
 }

--- a/transpiler/call.go
+++ b/transpiler/call.go
@@ -210,7 +210,7 @@ func transpileCallExpr(n *ast.CallExpr, p *program.Program) (
 	// Example of result Go code:
 	// i += 4
 	// _ = buffer
-	if functionName == "_" && functionDef.Name == "free" && functionDef.Substitution == "_" {
+	if functionDef.Substitution == "_" {
 		var argName string
 		if v, ok := realArgs[0].(*goast.CallExpr); ok && len(realArgs) == 1 {
 			if vv, ok := v.Args[0].(*goast.Ident); ok && len(v.Args) == 1 {
@@ -240,11 +240,6 @@ func transpileCallExpr(n *ast.CallExpr, p *program.Program) (
 			preStmts = append(preStmts, devNull)
 			return nil, "", preStmts, postStmts, nil
 		}
-	}
-
-	if functionName == "_" {
-		errorMessage := fmt.Sprintf("function with name: %s , cannot be used in Go", functionName)
-		p.AddMessage(p.GenerateWarningMessage(errors.New(errorMessage), n))
 	}
 
 	return util.NewCallExpr(functionName, realArgs...),

--- a/transpiler/transpiler.go
+++ b/transpiler/transpiler.go
@@ -173,6 +173,8 @@ func transpileToExpr(node ast.Node, p *program.Program, exprIsStmt bool) (
 		expr, exprType, err = transpileCharacterLiteral(n), "char", nil
 
 	case *ast.CallExpr:
+		// for implementation function `free` see transpileToStmt,
+		// for all another case:
 		expr, exprType, preStmts, postStmts, err = transpileCallExpr(n, p)
 
 	case *ast.CompoundAssignOperator:
@@ -276,6 +278,72 @@ func transpileToStmt(node ast.Node, p *program.Program) (
 
 		stmt = &goast.EmptyStmt{}
 		return
+	}
+
+	// Injection for specific implementation of function `free`
+	if n, ok := node.(*ast.CallExpr); ok {
+		if len(n.Children()) == 2 {
+			if imp1, ok := n.Children()[0].(*ast.ImplicitCastExpr); ok {
+				if decl1, ok := imp1.Children()[0].(*ast.DeclRefExpr); ok {
+					functionName := decl1.Name
+					// For function `free` from lib `stdlib.h`
+					if functionName == "free" {
+						if imp2, ok := n.Children()[1].(*ast.ImplicitCastExpr); ok {
+							if imp3, ok := imp2.Children()[0].(*ast.ImplicitCastExpr); ok {
+								if decl, ok := imp3.Children()[0].(*ast.DeclRefExpr); ok {
+									/* Example of input:
+									   |-CallExpr 0x36cc830 <line:8:9, col:18> 'void'
+									   | |-ImplicitCastExpr 0x36cc818 <col:9> 'void (*)(void *)' <FunctionToPointerDecay>
+									   | | `-DeclRefExpr 0x36cc7a0 <col:9> 'void (void *)' Function 0x36e0570 'free' 'void (void *)'
+									   | `-ImplicitCastExpr 0x36cc878 <col:14> 'void *' <BitCast>
+									   |   `-ImplicitCastExpr 0x36cc860 <col:14> 'int *' <LValueToRValue>
+									   |     `-DeclRefExpr 0x36cc7c8 <col:14> 'int *' lvalue Var 0x36cc088 'buf1' 'int *'
+									*/
+									/*
+										Example of output:
+										_ = buf1
+										Example of AST tree:
+										2: *ast.AssignStmt {
+										.  Lhs: []ast.Expr (len = 1) {
+										.  .  0: *ast.Ident {
+										.  .  .  NamePos: 10:1
+										.  .  .  Name: "_"
+										.  .  }
+										.  }
+										.  TokPos: 10:3
+										.  Tok: =
+										.  Rhs: []ast.Expr (len = 1) {
+										.  .  0: *ast.Ident {
+										.  .  .  NamePos: 10:5
+										.  .  .  Name: "d"
+										.  .  .  Obj: *(obj @ 78)
+										.  .  }
+										.  }
+										}
+									*/
+									devNull := &goast.AssignStmt{
+										Lhs: []goast.Expr{
+											&goast.Ident{
+												Name: "_",
+											},
+										},
+										Tok: token.ASSIGN,
+										Rhs: []goast.Expr{
+											&goast.Ident{
+												Name: decl.Name,
+											},
+										},
+									}
+									stmt = devNull
+									err = nil
+									return
+								}
+							}
+						}
+					}
+				}
+			}
+		}
 	}
 
 	// We do not care about the return type.

--- a/transpiler/transpiler.go
+++ b/transpiler/transpiler.go
@@ -173,8 +173,6 @@ func transpileToExpr(node ast.Node, p *program.Program, exprIsStmt bool) (
 		expr, exprType, err = transpileCharacterLiteral(n), "char", nil
 
 	case *ast.CallExpr:
-		// for implementation function `free` see transpileToStmt,
-		// for all another case:
 		expr, exprType, preStmts, postStmts, err = transpileCallExpr(n, p)
 
 	case *ast.CompoundAssignOperator:
@@ -209,7 +207,14 @@ func transpileToStmts(node ast.Node, p *program.Program) ([]goast.Stmt, error) {
 	}
 
 	stmt, preStmts, postStmts, err := transpileToStmt(node, p)
-	stmts := append(preStmts, stmt)
+	var stmts []goast.Stmt
+	// nil is happen, when we remove function `free` of <stdlib.h>
+	// see function CallExpr in transpiler
+	if stmt != nil {
+		stmts = append(preStmts, stmt)
+	} else {
+		stmts = preStmts
+	}
 	stmts = append(stmts, postStmts...)
 	return stmts, err
 }
@@ -280,79 +285,17 @@ func transpileToStmt(node ast.Node, p *program.Program) (
 		return
 	}
 
-	// Injection for specific implementation of function `free`
-	if n, ok := node.(*ast.CallExpr); ok {
-		if len(n.Children()) == 2 {
-			if imp1, ok := n.Children()[0].(*ast.ImplicitCastExpr); ok {
-				if decl1, ok := imp1.Children()[0].(*ast.DeclRefExpr); ok {
-					functionName := decl1.Name
-					// For function `free` from lib `stdlib.h`
-					if functionName == "free" {
-						if imp2, ok := n.Children()[1].(*ast.ImplicitCastExpr); ok {
-							if imp3, ok := imp2.Children()[0].(*ast.ImplicitCastExpr); ok {
-								if decl, ok := imp3.Children()[0].(*ast.DeclRefExpr); ok {
-									/* Example of input:
-									   |-CallExpr 0x36cc830 <line:8:9, col:18> 'void'
-									   | |-ImplicitCastExpr 0x36cc818 <col:9> 'void (*)(void *)' <FunctionToPointerDecay>
-									   | | `-DeclRefExpr 0x36cc7a0 <col:9> 'void (void *)' Function 0x36e0570 'free' 'void (void *)'
-									   | `-ImplicitCastExpr 0x36cc878 <col:14> 'void *' <BitCast>
-									   |   `-ImplicitCastExpr 0x36cc860 <col:14> 'int *' <LValueToRValue>
-									   |     `-DeclRefExpr 0x36cc7c8 <col:14> 'int *' lvalue Var 0x36cc088 'buf1' 'int *'
-									*/
-									/*
-										Example of output:
-										_ = buf1
-										Example of AST tree:
-										2: *ast.AssignStmt {
-										.  Lhs: []ast.Expr (len = 1) {
-										.  .  0: *ast.Ident {
-										.  .  .  NamePos: 10:1
-										.  .  .  Name: "_"
-										.  .  }
-										.  }
-										.  TokPos: 10:3
-										.  Tok: =
-										.  Rhs: []ast.Expr (len = 1) {
-										.  .  0: *ast.Ident {
-										.  .  .  NamePos: 10:5
-										.  .  .  Name: "d"
-										.  .  .  Obj: *(obj @ 78)
-										.  .  }
-										.  }
-										}
-									*/
-									devNull := &goast.AssignStmt{
-										Lhs: []goast.Expr{
-											&goast.Ident{
-												Name: "_",
-											},
-										},
-										Tok: token.ASSIGN,
-										Rhs: []goast.Expr{
-											&goast.Ident{
-												Name: decl.Name,
-											},
-										},
-									}
-									stmt = devNull
-									err = nil
-									return
-								}
-							}
-						}
-					}
-				}
-			}
-		}
-	}
-
 	// We do not care about the return type.
 	expr, _, preStmts, postStmts, err = transpileToExpr(node, p, true)
 	if err != nil {
 		return
 	}
 
-	stmt = util.NewExprStmt(expr)
+	// nil is happen, when we remove function `free` of <stdlib.h>
+	// see function CallExpr in transpiler
+	if expr != (*goast.CallExpr)(nil) {
+		stmt = util.NewExprStmt(expr)
+	}
 
 	return
 }

